From 34d71abcc34b80ae84f51f7ffef301e53c469ca0 Mon Sep 17 00:00:00 2001
From: Howard Hsu <howard-yh.hsu@mediatek.com>
Date: Mon, 8 Jun 2026 16:50:03 +0800
Subject: [PATCH] mtk: mac80211: fix mld sta could not join multi-link btwt
 schedule

fixup! mtk: mac80211: add support twt setup for multi-link operation

In process_btwt_id_0(), twt->params is shared across all link
iterations but not cleared before each drv_add_twt_setup() call.
Stale data from a previous link's setup may corrupt the params for
subsequent links, preventing an MLD station from successfully
joining a broadcast TWT schedule on all active links.

Fix this by zeroing twt->params at the beginning of each link
iteration.

Signed-off-by: Howard Hsu <howard-yh.hsu@mediatek.com>
---
 net/mac80211/cfg.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/net/mac80211/cfg.c b/net/mac80211/cfg.c
index 73d11d2e..8fcf6cae 100644
--- a/net/mac80211/cfg.c
+++ b/net/mac80211/cfg.c
@@ -1950,6 +1950,7 @@ static int process_btwt_id_0(struct ieee80211_local *local,
 {
 	struct ieee80211_link_sta *link_sta;
 	struct ieee80211_twt_setup *twt;
+	struct ieee80211_twt_params *params;
 	unsigned int link_id;
 	int ret;
 
@@ -1961,11 +1962,14 @@ static int process_btwt_id_0(struct ieee80211_local *local,
 		return -ENOMEM;
 
 	twt->control |= IEEE80211_TWT_NEGO_TYPE_BTWT_MBR_MGMT << 2;
+	params = (struct ieee80211_twt_params *)twt->params;
 
 	for_each_sta_active_link(&sdata->vif, &sta->sta, link_sta, link_id) {
 		ret = drv_add_twt_setup(local, sdata, &sta->sta, link_id, twt);
 		if (ret)
 			break;
+
+		memset(params, 0, sizeof(*params));
 	}
 	kfree(twt);
 
-- 
2.45.2

